test_tablets: add rack decommission test cases

bhalevy opened the pull request:
test_tablets: add rack decommission test cases

Test scenarios where decommissioing a compelte rack
should succeed, and reproduce scylladb/scylladb#19475
where decommissioning a rack would fail since the
number of remaining racks is insufficient to satisfy
the replication factor, even though the number of nodes
is sufficient, enshrining this behavior.

Refs scylladb/scylladb#19475

* This PR adds unit tests and improves an error message.  No backport required.

bhalevy commented:
Draft until https://github.com/scylladb/scylladb/pull/20730

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11805/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11805/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/719/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11787/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3773/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11805/)
#### Failed Tests (1/35425):
- [test_decommission_rack_after_adding_new_rack.release.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11805/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_decommission_rack_after_adding_new_rack_release_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_after_adding_new_rack.release.4)
#### Build Details:
- Duration: 6 hr 51 min
- Builder: i-08b0d91db89317f93 (m5ad.8xlarge)

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+                assert i == 0
+            else:
+                # last remaining node in this rack should contain all replicas in the rack
+                abs(i - expected_replicas_per_shard * nodes_per_rack) <= 1
```
no assert here

bhalevy replied:
fixed in next version by generalizing verification using a pre-computed map of expected tablets per node

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+    rf = num_racks - 1
+    ks = "test"
+    table = "test"
+    for rack in range(1, num_racks + 1):
```
This looks like a good utility function `servers = await create_racks(racks_count, nodes_per_rack, property_file)`

bhalevy replied:
I can add it in this source file.  Not sure it is fit for a general purpose utility function.
Also the property file would be generated by the function, not passed to it.
We can pass the dc name (with e.g. `dc1` as default value)

pehala replied:
> Not sure it is fit for a general purpose utility function.

I suspect we will have more testcases where we need to setup entire rack in the future so I can see how that function would be useful

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+            await manager.decommission_node(s.server_id)
+
+    expected_replicas_per_shard = (initial_tablets * rf) / ((num_racks-1) * nodes_per_rack) / shards_count
+    tablet_count = await get_tablet_count_per_shard_for_hosts(shards_count, manager, list(servers.values()), tables)
```
why are you converting `servers.values()` to list? You do not need list, you should only need iterable and that should be satisfiable by `servers.values()`

bhalevy replied:
okay, I'll test that

bhalevy replied:
not sure how to type an iterable as a parameter, so I left it as is in the next version.

pehala replied:
`collections.abc import Iterable`

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+                assert abs(i - expected_replicas_per_shard) <= 1
+            else:
+                assert i ==0
+    assert total == initial_tablets * rf
```
This test uses a lot of asserts, how about adding error messages to them, to make it more clear from the error failure what went wrong.

bhalevy replied:
The error quotes the expression that failed and the actual values.
I hope that the code is clear enough to describe the requirement.
An error message would basically repeat the expression

bhalevy replied:
this is addressed in the next version (I hope), by defining a helper function that creates a map of the expected tablet count per node.  I hope the calculation is clearer now.

pehala replied:
It is!

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+    await cql.run_async(f"CREATE KEYSPACE test WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': {rf}}} AND tablets = {{'initial': {initial_tablets}}};")
+    await cql.run_async(f"CREATE TABLE {ks}.{table} (pk int PRIMARY KEY, c int);")
+
+    keys = range(100)
```
keys should be a number not a range, this results in you having to call len on it in the future

bhalevy replied:
IIRC I copy-pasted it from an existing test, but that makes sense

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+
+    logger.info("Populate data")
+    cql = manager.get_cql()
+    await cql.run_async(f"CREATE KEYSPACE test WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': {rf}}} AND tablets = {{'initial': {initial_tablets}}};")
```
Another possible utility function: `ks, table = create_table(rf, initial_tablets)`, that way we even get rid of the static at the beggining of each test
```
    ks = "test"
    table = "test"
```

Sidenote: if initail_tablets are not important for this test, we could set it as a defaut in the util function and that would be one more static line hat would not need to present for each test

bhalevy replied:
The test needs to know it for the assertions below, so I don't want it to double guess the default value, but rather pass the arbitrary value to the create function so we have one authoritative value.

pehala replied:
Test does need to know it, but it doesnt care about its value, it is not important for the testcase, so an utility function or fixture can be used and it can return the `ks` and `table` with some value, which we can use in the test

bhalevy replied:
done in next version

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+            logger.debug(f"Decommissioning server={s}")
+            await manager.decommission_node(s.server_id)
+
+    expected_replicas_per_shard = (initial_tablets * rf) / ((num_racks-1) * nodes_per_rack) / shards_count
```
Do we need double division here or could this be simplified into one?
Sidenote: this equation appears in all of the tests, maybe it would be good to extract it somewhere else?

bhalevy replied:
Maybe I can refactor it to a function that will get higher level parameters that will make the logic clearer

bhalevy replied:
simplified in next version by more elaborate logic that explains the algorithm

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+
+    decommision_rack = f"rack{num_racks}"
+    decommision_count = 0
+    decommisioned_nodes = set[ServerNum]()
```
This is very interesting construct, I have never seen typing used like this. I would expect it to look like:
`decommisioned_nodes: set[ServerNum] = set()`
I could not find this expression in https://peps.python.org/pep-0526/, is something new I missed?

bhalevy replied:
I just used it naively as it made sense to me.
I don't mind using the "standard" notation of `decommisioned_nodes: set[ServerNum] = set()`

bhalevy replied:
made conservative in next version (though I like the alternative, non-standard notion better BTW)

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+
+    logger.info("Add a new rack")
+    new_rack = f"rack{num_racks}"
+    rack_servers = await manager.servers_add(nodes_per_rack, property_file={"dc": "dc1", "rack": new_rack})
```
This looks like a good utility function `rack_servers = await create_rack(nodes_per_rack, property_file)`

bhalevy replied:
I'll try to refactor it to a helper that construct the cluster given the number of datacenters and racks in each datacenter.

bhalevy replied:
done in next version

pehala reviewed (commented):
There is way too much copy paste code in the new tests (test_tablets.py), which makes it hard for me to orient myself and focus on whats actually important for the test (i.e. how the tests differ from each other) and whats just shared code. I think extracting common logic into utility functions and simplifying tests would go a long way here

bhalevy commented:
In  23e17e4:
- refactored tests and used utility functions to create the cluster, populate keyspace, table, and data, and to get a map of expected tablet counts, and verify that map against the actual tablet counts.

bhalevy commented:
rebased

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11998/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11998/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (4/1320):
- [test_decommission_rack_after_adding_new_rack.dev.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11998/testReport/junit/%28root%29/test_tablets/test_decommission_rack_after_adding_new_rack_dev_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_after_adding_new_rack.dev.10)
- [test_decommission_rack_after_adding_new_rack.dev.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11998/testReport/junit/%28root%29/test_tablets/test_decommission_rack_after_adding_new_rack_dev_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_after_adding_new_rack.dev.8)
- [test_decommission_rack_after_adding_new_rack.dev.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11998/testReport/junit/%28root%29/test_tablets/test_decommission_rack_after_adding_new_rack_dev_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_after_adding_new_rack.dev.9)
- [test_decommission_rack_after_adding_new_rack.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11998/testReport/junit/%28root%29/test_tablets/test_decommission_rack_after_adding_new_rack_release_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_after_adding_new_rack.release.6)
#### Build Details:
- Duration: 1 hr 10 min
- Builder: spider5.cloudius-systems.com

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1049,6 +1050,24 @@ async def get_tablet_count_per_shard_for_host(shards_count: int, manager: Manage
                         result[shard_id] += 1;
     return result
 
+async def get_tablet_count_per_shard_for_hosts(manager: ManagerClient, servers: list[ServerInfo], full_tables: dict[str, list[str]], shards_per_node: int = 2) -> dict[ServerNum, list[int]]:
```
Why this function cannot simply call `get_tablet_count_per_shard_for_host` for each of the hosts? Also I would appreciate if the order of arguments was the same, with only server being added as the last

bhalevy replied:
Because its complexity would be O(n^2) in the number of hosts.
If at all, I'd do it the other way around, where `get_tablet_count_per_shard_for_host` would call `get_tablet_count_per_shard_for_hosts` with a single server and manipulate the result to be backward compatible with the existing signature (for a single host)

bhalevy replied:
As for the parameters order, I moved `shards_per_node` to be last so it can be default-initialized.
Since the number of shards per host is not varied by the additional tests

bhalevy replied:
get_tablet_count_per_shard_for_host calls get_tablet_count_per_shard_for_hosts in next version

pehala replied:
> get_tablet_count_per_shard_for_host calls get_tablet_count_per_shard_for_hosts in next version

I would still like for them to have the same order of arguments as they as virtually the same, but we can leave it out for now and potentially change later. The fact that `get_tablet_count_per_shard_for_host` is just a "shortcut" is great improvement

bhalevy replied:
I reordered the parameters of `get_tablet_count_per_shard_for_host` in [268fcb5](https://github.com/scylladb/scylladb/commit/268fcb5f7cbd72782117dfc238552e9c0f509ba5):
```
async def get_tablet_count_per_shard_for_host(manager: ManagerClient, server: ServerInfo, full_tables: dict[str, list[str]], shards_count: int = 2) -> list[int]:
...
async def get_tablet_count_per_shard_for_hosts(manager: ManagerClient, servers: Iterable[ServerInfo], full_tables: dict[str, list[str]], shards_per_node: int = 2) -> dict[ServerNum, list[int]]:
```

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1296,3 +1315,186 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+    return servers
+
+
+async def create_and_populate_table(manager: ManagerClient, ks: str = "ks", table: str = "test", rf: int = 3, initial_tablets: int = 64, num_keys: int = 0) -> tuple[str, str, int, int]:
```
I would use None for default value for num_keys, it better indicates that the value is gonna be rewritten.

bhalevy replied:
okay. I actually coded it this way at first and it seemed overly engineered to me, but I can live with that.

bhalevy replied:
fixed in next version

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1296,3 +1315,186 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+
+    await asyncio.gather(*[cql.run_async(f"INSERT INTO {ks}.{table} (pk, c) VALUES ({k}, 1);") for k in range(num_keys)])
+
+    return (ks, table, initial_tablets, num_keys)
```
no need for the brackets

bhalevy replied:
well, they're parenthesis ;)
IMO it better represents that a tuple is being returned, but I speak Python only as a second (or third, forth, don't remember) language :)

pehala replied:
Yeah, it is a very minor nitpick, both are fine, without parentheses is little bit more common. Feel free to ignore

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1296,3 +1315,186 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+            live_servers[s.server_id] = s
+
+    logger.info("Verify tablet replicas distribution")
+    expected_replicas_per_server = get_expected_replicas_per_server(list(live_servers.values()), initial_tablets, rf)
```
These two lines are too long & verbose, it took me a minute to understand what is happening. It is also used a few times here (5times in the new tests), so I think it deserves yet another util function. Not sure how to name it:
`expected_replicas_per_server = get_expected_replicas_per_servers(live_servers, dead_servers, initial_tablets, rf)`

bhalevy replied:
I can just extend `get_expected_replicas_per_server` and the pre-decommission case can pass an empty list for dead_servers

pehala replied:
Good idea

pehala reviewed (commented):
I really like how you refactored the tests, it is much better now. Thanks! The rest of my comments are nitpicks

bhalevy commented:
In [0d2162a](https://github.com/scylladb/scylladb/commit/0d2162a43f08c2d7c1f1614f42e47ece99e97992):
- `get_tablet_count_per_shard_for_host`: use `get_tablet_count_per_shard_for_hosts`
- `get_expected_replicas_per_server`: added `dead_servers` param to simplify calculation after decommission (and before it, after adding a new rack that is expected to have no tablets, when load-balancing is disabled)
- `create_and_populate_table`: set num_keys default value to `None`
  - return tuple without surrounding it with parentheses
- `test_decommission_rack_after_adding_new_rack`
  - disabled tablet load-balancing around adding the new rack, otherwise load-balancer starts migrating tablets to it before decommission happens (making verification harder)

pehala commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -1296,3 +1306,191 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+            result[id] = replicas_per_node
+    for s in dead_servers:
+        result[s.server_id] = 0
+    #logger.debug(f"get_expected_replicas_per_server: live_servers={live_servers} dead_servers={dead_servers} initial_tablets={initial_tablets} rf={rf}: total_replicas={total_replicas} nodes_per_rack={nodes_per_rack}")
```
This should be either removed or uncommented

bhalevy replied:
done in next version

bhalevy commented:
In [268fcb5](https://github.com/scylladb/scylladb/commit/268fcb5f7cbd72782117dfc238552e9c0f509ba5):
- Used `Iterable` so that dictionary values can be passed directly to utility functions without listifying them (thanks @pehala for pushing on this)
- Added "topology_experimental_raft/test_tablets: get_tablet_count_per_shard_for_host: move shards_count param to be last"
  - small cleanup before next patch
- Cleaned up "test_tablets: add rack decommission test cases" to reduce the code churn

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12017/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12017/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (1/3960):
- [test_tablet_drain_failure_during_decommission.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12017/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.2)
#### Build Details:
- Duration: 2 hr 32 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12017/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12017/) **The following new/updated tests ran 100 times for each mode:** 🔹 auth_cluster/test_auth_no_quorum 🔹 auth_cluster/test_auth_v2_migration 🔹 auth_cluster/test_raft_service_levels 🔹 topology/test_automatic_cleanup 🔹 topology/test_change_ip 🔹 topology_custom/test_no_removed_node_event_on_ip_change 🔹 topology_custom/test_topology_failure_recovery 🔹 topology_experimental_raft/test_mv_tablets_replace 🔹 topology_experimental_raft/test_tablets 🔹 topology_experimental_raft/test_tablets_intranode
> 
> #### Failed Tests (1/3960):
> * [test_tablet_drain_failure_during_decommission.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12017/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_tablet_drain_failure_during_decommission.dev.2)
> 
> #### Build Details:
> * Duration: 2 hr 32 min
> * Builder: spider2.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/20060
respinning

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12032/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12032/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (1/1320):
- [test_tablet_drain_failure_during_decommission.dev.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12032/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_dev_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.9)
#### Build Details:
- Duration: 3 hr 31 min
- Builder: i-0dd49756ed017d42b (m5ad.8xlarge)

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12032/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12032/) **The following new/updated tests ran 100 times for each mode:** 🔹 auth_cluster/test_auth_no_quorum 🔹 auth_cluster/test_auth_v2_migration 🔹 auth_cluster/test_raft_service_levels 🔹 topology/test_automatic_cleanup 🔹 topology/test_change_ip 🔹 topology_custom/test_no_removed_node_event_on_ip_change 🔹 topology_custom/test_topology_failure_recovery 🔹 topology_experimental_raft/test_mv_tablets_replace 🔹 topology_experimental_raft/test_tablets 🔹 topology_experimental_raft/test_tablets_intranode
> 
> #### Failed Tests (1/1320):
> * [test_tablet_drain_failure_during_decommission.dev.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12032/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_dev_9) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_tablet_drain_failure_during_decommission.dev.9)
> 
> #### Build Details:
> * Duration: 3 hr 31 min
> * Builder: i-0dd49756ed017d42b (m5ad.8xlarge)

Hit unrelated issue: https://github.com/scylladb/scylladb/issues/20060
Respinned https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12053/

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12053/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12053/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (1/10560):
- [test_tablet_drain_failure_during_decommission.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12053/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.debug.6)
#### Build Details:
- Duration: 14 hr
- Builder: i-0e6c723cb4ed8712d (m5ad.12xlarge)

bhalevy commented:
In v4 ([bb42248](https://github.com/scylladb/scylladb/commit/bb42248932e446a5b74cee0f56dc72c4f5685a5d))
- rebased

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12361/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12361/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (1/2820):
- [test_tablet_drain_failure_during_decommission.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12361/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.debug.3)
#### Build Details:
- Duration: 1 hr 59 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12361/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12361/) **The following new/updated tests ran 100 times for each mode:** 🔹 auth_cluster/test_auth_no_quorum 🔹 auth_cluster/test_auth_v2_migration 🔹 auth_cluster/test_raft_service_levels 🔹 topology/test_automatic_cleanup 🔹 topology/test_change_ip 🔹 topology_custom/test_no_removed_node_event_on_ip_change 🔹 topology_custom/test_topology_failure_recovery 🔹 topology_experimental_raft/test_mv_tablets_replace 🔹 topology_experimental_raft/test_tablets 🔹 topology_experimental_raft/test_tablets_intranode
> 
> #### Failed Tests (1/2820):
> * [test_tablet_drain_failure_during_decommission.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12361/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_tablet_drain_failure_during_decommission.debug.3)

Unrelated: https://github.com/scylladb/scylladb/issues/20060

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12388/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12388/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (1/1410):
- [test_decommission_not_enough_racks.release.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12388/testReport/junit/%28root%29/test_tablets/test_decommission_not_enough_racks_release_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_not_enough_racks.release.8)
#### Build Details:
- Duration: 3 hr 41 min
- Builder: i-03f5294b7dca91a2c (m5ad.8xlarge)

bhalevy commented:
In v5 ([89c236a](https://github.com/scylladb/scylladb/commit/89c236a377f8da1d2fa847321086a24ae33c8032)):
- used `CREATE KEYSPACE IF NOT EXISTS` to deflake the new tests

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12427/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12427/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (1/4230):
- [test_tablet_drain_failure_during_decommission.dev.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12427/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_dev_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.6)
#### Build Details:
- Duration: 3 hr 2 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12427/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12427/) **The following new/updated tests ran 100 times for each mode:** 🔹 auth_cluster/test_auth_no_quorum 🔹 auth_cluster/test_auth_v2_migration 🔹 auth_cluster/test_raft_service_levels 🔹 topology/test_automatic_cleanup 🔹 topology/test_change_ip 🔹 topology_custom/test_no_removed_node_event_on_ip_change 🔹 topology_custom/test_topology_failure_recovery 🔹 topology_experimental_raft/test_mv_tablets_replace 🔹 topology_experimental_raft/test_tablets 🔹 topology_experimental_raft/test_tablets_intranode
> 
> #### Failed Tests (1/4230):
> * [test_tablet_drain_failure_during_decommission.dev.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12427/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_dev_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.6)

#20060 again.
I'm not sure why it happens so frequently in this branch.

avikivity commented:
How long do those tests take to run?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12460/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12460/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (1/5640):
- [test_tablet_drain_failure_during_decommission.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12460/testReport/junit/%28root%29/test_topology_failure_recovery/test_tablet_drain_failure_during_decommission_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.debug.3)
#### Build Details:
- Duration: 9 hr 20 min
- Builder: i-00b948f39fb9b41eb (m5ad.8xlarge)

pehala commented:
> I'm not sure why it happens so frequently in this branch.

I think it is because you change the file (removed the import) and thus it is running 100times. When I run `topology_custom/test_topology_failure_recovery ` locally even with repeat=20, I can easily reproduce this error.

bhalevy commented:
> How long do those tests take to run?

@avikivity when running these test locally in isolation (in dev mode):

| test case | time |
| --- | --- |
| test_decommission_rack_basic | 0m19.656s |
| test_decommission_rack_after_adding_new_rack | 0m23.864s |
| test_decommission_not_enough_racks | 0m19.670s |

bhalevy commented:
In [c81584c](https://github.com/scylladb/scylladb/commit/c81584cf416257b8cf548e3db6041d6876af408c):
- rebased on top of https://github.com/scylladb/scylladb/commit/78ceaeabcab6902da80f238a5f982ebc00f3bd68

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12933/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12933/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Build Details:
- Duration: 20 hr
- Builder: i-0df94d1a992cb4ddd (m5ad.8xlarge)

bhalevy commented:
> #### 🔴 CI State: ABORTED
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12933/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12933/) **The following new/updated tests ran 100 times for each mode:** 🔹 auth_cluster/test_auth_no_quorum 🔹 auth_cluster/test_auth_v2_migration 🔹 auth_cluster/test_raft_service_levels 🔹 topology/test_automatic_cleanup 🔹 topology/test_change_ip 🔹 topology_custom/test_no_removed_node_event_on_ip_change 🔹 topology_custom/test_topology_failure_recovery 🔹 topology_experimental_raft/test_mv_tablets_replace 🔹 topology_experimental_raft/test_tablets 🔹 topology_experimental_raft/test_tablets_intranode

timed out.  respinned (just unit tests)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12942/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12942/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
#### Failed Tests (1/1410):
- [test_decommission_rack_basic.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12942/testReport/junit/%28root%29/test_tablets/test_decommission_rack_basic_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_rack_basic.dev.3)
#### Build Details:
- Duration: 3 hr 40 min
- Builder: i-0c52e149771067b3c (m5ad.8xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12954/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12954/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/82/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1527/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4601/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12627/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12954/)
#### Build Details:
- Duration: 11 hr
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12992/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12992/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_no_quorum
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: topology/test_automatic_cleanup
:small_blue_diamond: topology/test_change_ip
:small_blue_diamond: topology_custom/test_no_removed_node_event_on_ip_change
:small_blue_diamond: topology_custom/test_topology_failure_recovery
:small_blue_diamond: topology_experimental_raft/test_mv_tablets_replace
:small_blue_diamond: topology_experimental_raft/test_tablets
:small_blue_diamond: topology_experimental_raft/test_tablets_intranode
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/109/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4637/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1563/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12665/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12992/)
#### Build Details:
- Duration: 11 hr
- Builder: spider6.cloudius-systems.com
